### PR TITLE
fix for DNS resolution issues

### LINKED ISSUE #2161: Akka.Remote.HeliosTransport fails to bind on  IPV6; DNS under some circumstances
Related to https://github.com/akkadotnet/akka.net/issues/1971

Based on some user reports from 1.1 we know we have some regressions that affected DNS binding.

Reproduction spec report thus far:

image

### PR DESCRIPTION
**Pull request by Aaronontheweb**
Resolves #2161

upgraded to Helios 2.1.2; set default DNS resolution target to IPV4; made DNS resolution configurable via HOCON.

### CODE REVIEW

File: src/core/Akka.Remote.Tests/Transport/HeliosTransportDnsResolutionSpec.cs [hunk 1]
```diff
@@ -70,12 +69,13 @@ public HeliosTransportDnsResolutionSpec()
             Arb.Register(typeof(EndpointGenerators));
         }
 
-        public Config BuildConfig(string hostname, int? port = null, string publichostname = null)
+        public Config BuildConfig(string hostname, int? port = null, string publichostname = null, bool useIpv6 = false)
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote.Tests/Transport/HeliosTransportDnsResolutionSpec.cs]: Modified to use configurable IP version with DNS resolution

File: src/core/Akka.Remote.Tests/Transport/HeliosTransportDnsResolutionSpec.cs [hunk 2]
```diff
@@ -70,12 +69,13 @@ public HeliosTransportDnsResolutionSpec()
...
         {
             return ConfigurationFactory.ParseString(@"akka.actor.provider = ""Akka.Remote.RemoteActorRefProvider, Akka.Remote""")
                 .WithFallback("akka.remote.helios.tcp.hostname =\"" + hostname + "\"")
                 .WithFallback("akka.remote.helios.tcp.public-hostname =\"" + (publichostname ?? hostname) + "\"")
                 .WithFallback("akka.remote.helios.tcp.port = " + (port ?? 0))
+                .WithFallback("akka.remote.helios.tcp.dns-use-ipv6 = " + useIpv6.ToString().ToLowerInvariant())
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote.Tests/Transport/HeliosTransportDnsResolutionSpec.cs]: DNS resolution switch from configuration

File: src/core/Akka.Remote.Tests/Transport/HeliosTransportDnsResolutionSpec.cs [hunk 3]
```diff
@@ -158,6 +166,56 @@ public Property HeliosTransport_Should_Resolve_DNS(EndPoint inbound, EndPoint ou
             }
         }
 
+        [Property]
+        public Property HeliosTransport_Should_Resolve_DNS_with_PublicHostname(IPEndPoint inbound, DnsEndPoint publicInbound,
+            IPEndPoint outbound, DnsEndPoint publicOutbound, bool dnsUseIpv6)
+        {
+            if (dnsUseIpv6 &&
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote.Tests/Transport/HeliosTransportDnsResolutionSpec.cs]: Filter out invalid combinations as pre-conditions for the spec. Requires the end-user to be consistent about whether they're using IPV4 or IPV6 for DNS resolution.

File: src/core/Akka.Remote.Tests/Transport/HeliosTransportDnsResolutionSpec.cs [hunk 4]
```diff
@@ -158,6 +166,56 @@ public Property HeliosTransport_Should_Resolve_DNS(EndPoint inbound, EndPoint ou
             }
         }
 
+        [Property]
+        public Property HeliosTransport_Should_Resolve_DNS_with_PublicHostname(IPEndPoint inbound, DnsEndPoint publicInbound,
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote.Tests/Transport/HeliosTransportDnsResolutionSpec.cs]: Model-based test to verify all combinations of actual IP address + public hostname + DNS resolution preference.

File: src/core/Akka.Remote/Configuration/Remote.conf [hunk 5]
```diff
@@ -342,6 +342,10 @@ akka {
 	  # this is designed to make it easy to support private / public addressing schemes
 	  public-hostname = ""
 
+	  # If set to true, we will use IPV6 addresses upon DNS resolution for host names.
+	  # Otherwise, we will use IPV4.
+	  dns-use-ipv6 = false
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/Configuration/Remote.conf]: New setting added to Remote.config - sets the default resolution strategy to IPV4, which restores backwards compatibility with previous versions of Akka.NET.

File: src/core/Akka.Remote/Transport/Helios/HeliosTransport.cs [hunk 6]
```diff
@@ -287,6 +290,7 @@ protected ClientBootstrap ClientFactory(Address remoteAddres)
                     .Option(ChannelOption.TcpNodelay, Settings.TcpNoDelay)
                     .Option(ChannelOption.ConnectTimeout, Settings.ConnectTimeout)
                     .Option(ChannelOption.AutoRead, false)
+                    .PreferredDnsResolutionFamily(Settings.DnsUseIpv6 ? AddressFamily.InterNetworkV6 : AddressFamily.InterNetwork)
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/Transport/Helios/HeliosTransport.cs]: Tells Helios which address family to use when resolving on the server

File: src/core/Akka.Remote/Transport/Helios/HeliosTransport.cs [hunk 7]
```diff
@@ -325,6 +329,7 @@ protected ServerBootstrap ServerFactory
                      .ChildOption(ChannelOption.TcpNodelay, Settings.TcpNoDelay)
                      .ChildOption(ChannelOption.AutoRead, false)
                      .Option(ChannelOption.SoBacklog, Settings.Backlog)
+                     .PreferredDnsResolutionFamily(Settings.DnsUseIpv6 ? AddressFamily.InterNetworkV6 : AddressFamily.InterNetwork)
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/Transport/Helios/HeliosTransport.cs]: Tells Helios which address family to use when resolving on the client

**Comment by Silv3rcircl3** [comment]
LGTM

**Pull request merged**